### PR TITLE
Jmx shards info docs

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

This is a draft for exposing additional node information using jmx.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/admin/monitoring.rst [hunk 1]
```diff
@@ -168,22 +168,43 @@ NodeStatus can be accessed using the JMX MBean object name
...
 
-   Provides the human friendly name of the node
++------------------+------------------------------------------------------+
+| Name             | Description                                          |
++==================+======================================================+
+| ``id``           | The id of the shard.                                 |
++------------------+------------------------------------------------------+
+| ``size``         | The size in bytes of the shard.                      |
++------------------+------------------------------------------------------+
+| ``status``       | The status of the shard.                             |
```
Comment (mfussenegger) [inline docs/admin/monitoring.rst:206]: What kind of states are there?

In which format are we going to expose this to prometheus in the jmx exporter? Is the state going to be a propery that can be filtered on?

**Comment by mkleen** [comment]
i will do another iteration.

File: docs/admin/monitoring.rst [hunk 2]
```diff
@@ -168,22 +168,43 @@ NodeStatus can be accessed using the JMX MBean object name
...
+The ``NodeInfo`` MXBean exposes information of a ``CrateDB`` node.
 
-NodeInfo can be accessed using the JMX MBean object name
-``io.crate.monitoring:type=NodeInfo`` and the following attributes:
-
- - ``ClusterStateVersion``
-
-   Provides the version of the current applied cluster state
-
- - ``NodeId``
```
Comment (mfussenegger) [inline docs/admin/monitoring.rst:180]: Are we changing this to `id` ? Would be a breaking change

File: docs/admin/monitoring.rst [hunk 3]
```diff
@@ -168,22 +168,43 @@ NodeStatus can be accessed using the JMX MBean object name
 NodeInfo MBean
 --------------
 
-The ``NodeInfo`` JMX MBean exposes information about the current node;
+The ``NodeInfo`` MXBean exposes information of a ``CrateDB`` node.
 
-NodeInfo can be accessed using the JMX MBean object name
-``io.crate.monitoring:type=NodeInfo`` and the following attributes:
-
- - ``ClusterStateVersion``
```
Comment (mfussenegger) [inline docs/admin/monitoring.rst:176]: We should keep this attribute, no?

**Comment by mkleen** [comment]
sorry this was too early. i created initially a new mxbean, and just realized i could merge it with nodesinfo.

File: docs/admin/monitoring.rst [hunk 4]
```diff
@@ -173,17 +173,45 @@ The ``NodeInfo`` JMX MBean exposes information about the current node;
...
-
-   Provides the human friendly name of the node
++------------------+------------------------------------------------------+
+| Name             | Description                                          |
++==================+======================================================+
+| ``id``           | The id of the shard.                                 |
++------------------+------------------------------------------------------+
+| ``size``         | The size in bytes of the shard.                      |
++------------------+------------------------------------------------------+
+| ``state``        | The state of the shard, possible states are:         |
```
Comment (mkleen) [inline docs/admin/monitoring.rst:212]: You mentioned there should be an option to filter by shard state ?

Comment (mfussenegger) [reply docs/admin/monitoring.rst:212]: Well, it's more about thinking about how users are going to consume these metrics. We'll need to extend https://github.com/crate/jmx_exporter to also expose these metrics for prometheus and need to come up with a format that is useful for users who want to consume them.

File: docs/admin/monitoring.rst (hunk 1 above)
Comment (mkleen) [reply docs/admin/monitoring.rst]: My idea was to expose the shard related information in the following way to prometheus:

```
crate_node(name=shards_info, property=number_of_primary_shards) 1
crate_node(name=shards_info, property=number_of_replicas) 2
crate_node(name=shards_info, property=number_of_shards) 3
crate_node(name=shards_info, property=number_of_unassigned_shards) 0

crate_shards(id=1, property=size) 300
crate_shards(state=started) 1 
crate_shards(state=recovering) 1 
```

Comment (mfussenegger) [reply docs/admin/monitoring.rst]: Sounds good :+1:

**Pull request closed without merging**